Optimize performance (~4X faster build when images are up-to-date)

skovhus opened the pull request:
I've looked into Brick and wanted to figure out why building the tmrow repository was slower than expected. 

I got these timings locally when nothing changed and everything was already prepared and built:

    time brick -r prepare    : 48 seconds (same time if for example .npmrc was touched)
    time brick -r build      : 175 seconds

I would expect an almost instant prepare and build and time when the Docker images were up-to-date. 

It turns out that the overhead of running Buildit/Docker is rather slow for that many images – even though layers are cached. I tried to do some experiments with a plain Dockerfile and observed the same overhead.

I came up with two improvements some time ago... But now I managed to wrap it up into a PR. :) 

### UPDATE: Buildkit/Docker performance

The primary reasons for the performance issue with Buildkit on my machine are:
- general overhead in starting up Docker and running through each layer
- Slowest layers:  "resolve image config for" (0.4-1.2s), "[internal] load metadata for" (0.4s)



### Improvement 1: dependency change detection 

Idea: not running docker on `brick prepare` and `brick build` if input(s) and the BUILD.yaml file didn't change since the last image was tagged. I tried to think if this would work in all cases, and at least logically I cannot see why a build should be done if the inputs didn't change. Right? 🤔

UPDATE: originally I used the using file modified timestamps on input paths. But as file timestamps are not preserved in git, then CI wouldn't be able to cache anything as cloning the repository would update the modified timestamp. Instead we now hash all dependencies (inputs and BUILD.yaml file). The hashes are stored in a `~/.brick-cache.json` file.

When everything is prepared/built already:

    time brick -r prepare    : 1.5 seconds
    time brick -r build      : 66 seconds

So when everything is up-to-date: brick build time is 2.5X faster and prepare time is 32X faster. 🚀


### Improvement 2: output extraction using get_archive

The output extraction after a build is very slow. I noticed a comment around this to use `get_archive`. Doing so improves the performance a bit:

    time brick -r build      : 144 seconds (without improvement 1)


### Improvement 1 + 2:

When everything is already built:

    time brick -r build      : 46 seconds

So we end up with: brick build time is 4X faster and prepare time is still 24X faster. 🚀

But the build time should be instant when nothing changed, right? It turns out that collecting outputs is still very slow and is always done. 

When disabling any collecting of outputs:
    time brick -r build      : 5 seconds


### Conclusion

Of course these timing are made when everything is up-to-date. But as most PRs only touches a few images, then the performance improvement should be substantial on CI.

Next step could be to be more clever when doing output extraction using hashing. 🌤 But maybe that performance increase isn't worth the added complexities.


### TODO

- [x] Sanity check my assumption that prepare and build should not be done unless inputs or BUILD.yaml file changed
- [x] Performance test on CI (modified timestamp)
- [ ] Verify that we didn't break anything (mostly this would be stale builds, i.e. not building when we should)

skovhus committed:
Add support for skipping redundant docker build

Timing before any improvements when nothing changed and everything was prepared and built:
  time brick -r prepare    : 48 seconds (same time if .npmrc was touched)
  time brick -r build      : 175 seconds

When everything is prepared/built already:
  time brick -r prepare    : 2 seconds
  time brick -r build      : 66 seconds (as collect is rather slow and some targets rebuilds due to overlapping inputs and outputs)

skovhus committed:
Centralize config reading

skovhus committed:
Optimize performance by skipping redundant builds

skovhus committed:
Improve error message for expand_inputs

skovhus committed:
Speed up output extraction by using get_archive

This gives an performance improvement from 175 to 144 seconds on tmrow.

skovhus committed:
Remove outdated comment (nothing is named cache_from)

madsnedergaard commented:
This is epic, I've spent so much time trying to figure out why it's running so slowly in CI (but fine locally) 😭 

Will get straight to testing it now!

skovhus commented:
> This is epic, I've spent so much time trying to figure out why it's running so slowly in CI (but fine locally) 😭

Glad to hear! All timings here are local. I haven't tested on CI yet (and don't plan to do that before September).

So any help testing would be greatness! 👏

madsnedergaard commented on brick/dockerlib.py:
```diff
@@ -12,6 +13,36 @@
+    inspect_result = subprocess.run(
+        f"docker inspect -f '{{{{ json .Metadata.LastTagTime }}}}' {tag}",
+        shell=True,
+        capture_output=True)
```
The CI server is running Python 3.6, so `capture_output` is not supported (it was introduced in 3.7) - could we try with a workaround to avoid updating the server while testing? :)
([here's a few suggestions](https://stackoverflow.com/questions/53209127/subprocess-unexpected-keyword-argument-capture-output/53209196))

skovhus replied:
Should be fixed now. :)

skovhus committed:
Support Python 3.6 by removing capture_output

corradio commented:
This is quite a clever way to optimise the build. I'll let Mads review it but from my point of view it looks good to me!

madsnedergaard commented on brick/dockerlib.py:
```diff
@@ -12,6 +14,38 @@
 docker_client = docker.from_env()
 
 
+def is_built_up_to_date(tag: str, dependency_paths: List[str]) -> bool:
```
First up, thanks for actually doing this work now - and let me know if you'd prefer to wait with continuing work on it until you're actually in the office! 😄 

---

As far as I can tell, this way of checking for dependencies doesn't support listing full directories and using globbing - it only checks when the actual directory was modified and not the content 🤔 

Here's a few examples of how dependencies are being listed:

```bash
	inputs:
	  - src
      - ../../../{co2engine,utils}
      - ../../../../northapp/src/errorUtils.js
      - ../../../../northapp/src/schemas/
      - gatsby-*.js
      - ./*.py
      - ./{.babelrc,babel.config.js,.eslintignore}
```

Thirdly, is looping through dependencies and checking file/folder timestamps for each the best way of checking for changes? I would imagine it could get quite intense for some packages, although maybe it's speedy enough to not be a problem (fx if someone accidentally added `node_modules`)?

I remember previously looking into using `git` for listing the files changed in a branch, but can't remember why I stopped/never took it further 🙈 

For inspiration, [baur](https://github.com/simplesurance/baur) (another build tool) does it like this: 
> baur calculates a SHA384 digest for all inputs and stores the digest in the database when an application was built and its artifacts uploaded (baur build). The digest is used to detect if a previous build for the same input files exists. If a build exist, the application does not need to be rebuilt, otherwise a build is done. 

**EDIT:**
Another build tool (`mbt`) is also using `git` to diff, so there might be something for inspiration in there: https://github.com/mbtproject/mbt/blob/master/cmd/build.go

corradio replied:
> I remember previously looking into using git for listing the files changed in a branch, but can't remember why I stopped/never took it further 🙈

Jumping in here to refresh our memory: It's probably because you might want to re-build even though nothing changed in git (for example if the last build crashed).

skovhus replied:
> As far as I can tell, this way of checking for dependencies doesn't support listing full directories and using globbing - it only checks when the actual directory was modified and not the content 🤔

Globbing is supported here as the dependency_paths are already expanded. But I did just check, and any changes in a folder does not reflect the modified timestamp. So we need to recursively check folders for changes.

> Thirdly, is looping through dependencies and checking file/folder timestamps for each the best way of checking for changes? I would imagine it could get quite intense for some packages, although maybe it's speedy enough to not be a problem (fx if someone accidentally added node_modules)?

Accessing file information is really fast, so I don't think it should be an issue. But let us see once the recursive traversal of folders is implemented. :)

skovhus replied:
@madsnedergaard do you have some timings for trying this out on CI? And what about locally, do you see similar speed ups?

skovhus replied:
> First up, thanks for actually doing this work now - and let me know if you'd prefer to wait with continuing work on it until you're actually in the office! 😄

It was fun! And it was nice to wrap it up in a PR. Let us see how much time I have, else we will wrap it up at a later point. ;)

madsnedergaard replied:
Yeah, no change in timings - it's building every package.

But I guess it's because files are created from fresh in CI on each build, so file timestamps is an issue:

![Screenshot 2020-08-25 at 15 32 18](https://user-images.githubusercontent.com/3296643/91180591-2ffafb00-e6e8-11ea-8378-70635bd688cf.png)

https://drone.tmrow.com/tmrowco/tmrow/5082/1/3

skovhus replied:
> But I guess it's because files are created from fresh in CI on each build, so file timestamps is an issue:

Auch. I missed the fact that file time stamps are not preserved in Git. Then my approach here is flawed... There are script to update file time stamps based on git history, but I don’t like to rely on that. 

I’ll probably pull out the extract optimization into a separate PR. We could probably also do some more optimizations here. 

And we should debug why build kit is that slow at diffing.

corradio commented:
I'm still surprised that buildkit is not doing this by itself. Maybe worth opening up an issue on https://github.com/moby/buildkit/ or reaching out to the folks.

skovhus commented:
> I'm still surprised that buildkit is not doing this by itself. 

Yes, we should debug this a bit more. I observed a diffing overhead on a plain Dockerfile, but we could investigate further.

skovhus committed:
Use hash and cache file for dependency tracking

This is almost as fast as using the modified timestamp, and it does work on CI where git clone would update the modified timestamps.

skovhus commented:
Replaced by https://github.com/tmrowco/brick/pull/45 and https://github.com/tmrowco/brick/pull/46
